File: Client/Negwork.Client.App/scripts/app/categories/categories-service.js
(function () {
    'use strict';

    var categoriesService = function categoriesService(data) {
        var CATEGORIES_API_URL = '/api/categories';

        function getAll() {
            return data.get(CATEGORIES_API_URL);
        }

        function getById(id) {
            return data.get(CATEGORIES_API_URL + '/' + id);
        }

        function getForArticles() {
            return data.get(CATEGORIES_API_URL + '/articles');
        }

        function getForImages() {
            return data.get(CATEGORIES_API_URL + '/images');
        }

        function add(category) {
            return data.post(CATEGORIES_API_URL, category);
        }

        return {
            getAll: getAll,
            getById: getById,
            getForArticles: getForArticles,
            getForImages: getForImages,
            add: add
        };
    }

    angular
        .module('negwork.services')
        .factory('categories', ['data', categoriesService]);
}());

File: Services/Negwork.Services.Data/CategoriesService.cs
namespace Negwork.Services.Data
{
    using System.Linq;

    using Contracts;
    using Negwork.Data.Models;
    using Negwork.Data.Repositories;

    public class CategoriesService : ICategoriesService
    {
        private IRepository<Category> categories;

        public CategoriesService(IRepository<Category> categories)
        {
            this.categories = categories;
        }

        public IQueryable<Category> GetAll()
        {
            return this.categories.All();
        }

        public IQueryable<Category> GetArticleCategories()
        {
            return this.categories
                .All()
                .Where(c => c.Articles.Count > 0);
        }

        public IQueryable<Category> GetImageCategories()
        {
            return this.categories
                .All()
                .Where(c => c.Images.Count > 0);
        }

        public Category GetById(int id)
        {
            return this.categories.All()
                .Where(c => c.Id == id)
                .FirstOrDefault();
        }

        public Category CreateCategory(string name, string image)
        {
            var categoryWithName = this.categories
                .All()
                .Where(c => c.Name == name)
                .FirstOrDefault();

            if (categoryWithName != null)
            {
                return null;
            }

            var category = new Category()
            {
                Name = name,
                Image = image
            };

            this.categories.Add(category);
            this.categories.SaveChanges();

            return category;
        }
    }
}
